HPCC-24135 Can't start a cloud system without an esp

richardkchapman opened the pull request:
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

richardkchapman committed:
HPCC-24135 Can't start a cloud system without an esp

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman commented:
@jakesmith Please review

HPCCSmoketest commented:
Process of PR-13804, label: defaultEsp starts now.
The reason of this test is: Forced to re-test.
Commit ID: 6cdac10dabd4dbce6ce7a04c358bb3385971d482
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-82.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-82.ca-central-1.compute.internal
Sha: 6cdac10dabd4dbce6ce7a04c358bb3385971d482
Build: success
Milestone:Install hpccsystems-platform-community_7.8.19-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |54 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |42 sec  (00:00:42) |
|setup (roxie) |11 |11 |0 |18 sec  (00:00:18) |
|test (hthor) |874 |874 |0 |308 sec  (00:05:08) |
|test (thor) |812 |812 |0 |747 sec  (00:12:27) |
|test (roxie) |953 |953 |0 |391 sec  (00:06:31) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  651 sec (00:10:51) |  118 sec (00:01:58) |   22 sec (00:00:22) |   19 sec (00:00:19) | 1774 sec (00:29:34) |   18 sec (00:00:18) | 2611 sec (00:43:31) |

HPCCSmoketest commented:
Process of PR-13804, label: defaultEsp starts now.
The reason of this test is: Did not build yet.
Commit ID: 6cdac10dabd4dbce6ce7a04c358bb3385971d482
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 6cdac10dabd4dbce6ce7a04c358bb3385971d482
Build: success
Milestone:Install hpccsystems-platform-community_7.8.19-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |53 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |9 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |874 |874 |0 |578 sec  (00:09:38) |
|test (thor) |812 |812 |0 |1013 sec  (00:16:53) |
|test (roxie) |953 |953 |0 |710 sec  (00:11:50) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   11 sec (00:00:11) |  936 sec (00:15:36) |  122 sec (00:02:02) |   23 sec (00:00:23) |   17 sec (00:00:17) | 2654 sec (00:44:14) |   18 sec (00:00:18) | 3781 sec (01:03:01) |

jakesmith commented on dockerfiles/hpcc/templates/_helpers.tpl:
```diff
@@ -47,9 +47,17 @@ Generate global ConfigMap info
+ {{- $defaultEsp := .Values.global.defaultEsp -}}
+{{- else if hasPrefix "[]" (typeOf .Values.esp) -}}
+ {{- if (len .Values.esp) -}}
+  {{- $defaultEsp := (index .Values.esp 0).name -}}
```
I don't think this works, because of the scoping rules. It will set $defaultEsp in the scope of just this if block.
The trick I've seen to get this to work is:
```
{{- if (len .Values.esp) -}}
  $dummy := set . "defaultEsp"  (index .Values.esp 0).name -}}
{{- end }}

...

defaultEsp: {{ .defaultEsp }}
```

jakesmith commented on dockerfiles/hpcc/templates/_helpers.tpl:
```diff
@@ -47,9 +47,17 @@ Generate global ConfigMap info
+{{- if .Values.global.defaultEsp -}}
+ {{- $defaultEsp := .Values.global.defaultEsp -}}
+{{- else if hasPrefix "[]" (typeOf .Values.esp) -}}
+ {{- if (len .Values.esp) -}}
```
should also check is not disabled?

jakesmith commented:
@richardkchapman - couple of comments/questions

richardkchapman commented:
Hmmm, thought I had tried all combinations but clearly not!

I have reworked and tested a bit more. Also fixed the disabled flag (which was wrong in original implementation too) and the fact that you could not specify more than one esp in values.yaml

richardkchapman commented:
@jakesmith Please re-review

HPCCSmoketest commented:
Process of PR-13804, label: defaultEsp starts now.
The reason of this test is: New commit.
Commit ID: 17bf650a3d9e32bb435a39ed0c3dc5d387220b5f
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Process of PR-13804, label: defaultEsp starts now.
The reason of this test is: Forced to re-test.
Commit ID: 17bf650a3d9e32bb435a39ed0c3dc5d387220b5f
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-209.ca-central-1.compute.internal

jakesmith commented on dockerfiles/hpcc/templates/_helpers.tpl:
```diff
@@ -47,9 +47,25 @@ Generate global ConfigMap info
 singleNode: {{ .Values.global.singleNode | default false }}
-defaultEsp: {{ .Values.global.defaultEsp | default (index .Values.esp 0).name }}
+
+{{-/* Set the globalEsp to whatever was set globally, or the first non-disabled esp (if there is one) */-}}
```
I think needs a space after/before '-'

jakesmith replied:
@richardkchapman - I ran it and got 'parse error at (hpcc/templates/_helpers.tpl:54): "-"

jakesmith commented on dockerfiles/hpcc/templates/_helpers.tpl:
```diff
@@ -47,9 +47,25 @@ Generate global ConfigMap info
+  {{- end -}} 
+ {{- end -}} 
+{{- end }} 
+defaultEsp: {{ $local.defaultEsp | quote }}
```
should this be conditionally generated?, i.e. inside a
```
{{- if $local.defaultEsp }}
```

richardkchapman replied:
It's ok to set it to blank if not set

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-209.ca-central-1.compute.internal
Sha: 17bf650a3d9e32bb435a39ed0c3dc5d387220b5f
Build: success
Milestone:Install hpccsystems-platform-community_7.8.19-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |51 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |874 |874 |0 |306 sec  (00:05:06) |
|test (thor) |812 |812 |0 |744 sec  (00:12:24) |
|test (roxie) |953 |953 |0 |386 sec  (00:06:26) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  638 sec (00:10:38) |  116 sec (00:01:56) |   23 sec (00:00:23) |   20 sec (00:00:20) | 1766 sec (00:29:26) |   18 sec (00:00:18) | 2590 sec (00:43:10) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 17bf650a3d9e32bb435a39ed0c3dc5d387220b5f
Build: success
Milestone:Install hpccsystems-platform-community_7.8.19-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |9 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |874 |874 |0 |569 sec  (00:09:29) |
|test (thor) |812 |812 |0 |1028 sec  (00:17:08) |
|test (roxie) |953 |953 |0 |713 sec  (00:11:53) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  935 sec (00:15:35) |  123 sec (00:02:03) |   22 sec (00:00:22) |   17 sec (00:00:17) | 2641 sec (00:44:01) |   17 sec (00:00:17) | 3764 sec (01:02:44) |

HPCCSmoketest commented:
Process of PR-13804, label: defaultEsp starts now.
The reason of this test is: Forced to re-test.
Commit ID: 852fec7706f7ff4d5677ef21afb1f14e2a2f2e35
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-69.ca-central-1.compute.internal

jakesmith commented:
@richardkchapman - looks good to merge to me.

richardkchapman commented:
@ghalliday Please review/merge

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-69.ca-central-1.compute.internal
Sha: 852fec7706f7ff4d5677ef21afb1f14e2a2f2e35
Build: success
Milestone:Install hpccsystems-platform-community_7.8.19-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |32 sec  (00:00:32) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |874 |874 |0 |331 sec  (00:05:31) |
|test (thor) |812 |812 |0 |756 sec  (00:12:36) |
|test (roxie) |953 |953 |0 |440 sec  (00:07:20) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  695 sec (00:11:35) |  117 sec (00:01:57) |   23 sec (00:00:23) |   19 sec (00:00:19) | 1859 sec (00:30:59) |   19 sec (00:00:19) | 2741 sec (00:45:41) |

HPCCSmoketest commented:
Process of PR-13804, label: defaultEsp starts now.
The reason of this test is: New commit.
Commit ID: 852fec7706f7ff4d5677ef21afb1f14e2a2f2e35
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 852fec7706f7ff4d5677ef21afb1f14e2a2f2e35
Build: success
Milestone:Install hpccsystems-platform-community_7.8.19-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |54 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |9 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |874 |874 |0 |565 sec  (00:09:25) |
|test (thor) |812 |812 |0 |998 sec  (00:16:38) |
|test (roxie) |953 |953 |0 |704 sec  (00:11:44) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  905 sec (00:15:05) |  123 sec (00:02:03) |   25 sec (00:00:25) |   17 sec (00:00:17) | 2622 sec (00:43:42) |   19 sec (00:00:19) | 3720 sec (01:02:00) |
